editors/document-model/useSchemaEditor.ts

```diff
@@ -0,0 +1,70 @@
+import { State } from 'document-model/document';
+import {
+    DocumentModelState,
+    DocumentModelLocalState,
+    DocumentSpecification,
+} from 'document-model/document-model';
+import { useEffect, useState } from 'react';
+import z from 'zod';
+
+import { ScopeType } from './editor-schema';
+import { isJSONEqual } from '../common/json-editor';
+
+type SchemaState = {
+    documentName: string;
+    schema: string;
+    validator: () => z.AnyZodObject;
+};
+
+interface UseSchemaEditorProps {
+    scope: ScopeType;
+    state: State<DocumentModelState, DocumentModelLocalState>;
+    setInitialState: (initialValue: string, scope: ScopeType) => void;
+}
+
+export interface SchemaResult {
+    schemaState?: SchemaState;
+    setSchemaState: React.Dispatch<
+        React.SetStateAction<SchemaState | undefined>
+    >;
+    specification?: DocumentSpecification;
+    setInitialValue: React.Dispatch<React.SetStateAction<JSON>>;
+    initialValue: JSON;
+}
+
+export const useSchemaEditor = (props: UseSchemaEditorProps): SchemaResult => {
+    const { state, scope, setInitialState } = props;
+
```

Comment:
I think it's worth it to create a variable for`state[scope]` and `specification?.state[scope]` to make the rest of the code simpler.

Reply:
fixed: 13e334dc9004e062e5dd4930d922879a4704ca95